### PR TITLE
Integrate TE-FP8 to optimum-habana

### PR DESCRIPTION
**Pull request by SanjuCSudhakaran**

### CODE REVIEW

File: optimum/habana/accelerate/utils/transformer_engine.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+def is_fp8_available():
+    return has_transformer_engine
+
+
+def _convert_model(model, to_transformer_engine=True, _convert_linear=True):
+    """
+    Recursively converts the linear and layernorm layers of a model to their `transformers_engine` counterpart.
+    """
+    if not is_fp8_available():
```
Comment (vivekgoe) [inline optimum/habana/accelerate/utils/transformer_engine.py:38]: Lets add support for SDPA also, can be ported from MLPerf branch. You can default it to disabled by setting _convert_sdpa=False for now. Later we will figure out a way to enable it only for Gaudi3.

File: optimum/habana/accelerate/utils/transformer_engine.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        raise ImportError("Using `convert_model` requires transformer_engine to be installed.")
+    for name, module in model.named_children():
+        if type(module) == PEFTLinear and to_transformer_engine and _convert_linear:
+            LoRALinear.replace_forward(module)
+        if isinstance(module, torch.nn.Linear) and not type(module) == PEFTLinear and to_transformer_engine and _convert_linear:
```
Comment (vivekgoe) [inline optimum/habana/accelerate/utils/transformer_engine.py:43]: Do we enter this check for LoRALinear layer? If yes, why are we creating another te.Linear? We created one as part of LoRALinear init.

File: optimum/habana/accelerate/utils/transformer_engine.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+            has_bias = module.bias is not None
+            te_module = te.Linear(
+                module.in_features, module.out_features, bias=has_bias, params_dtype=module.weight.dtype, skip_weight_param_allocation=True
```
Comment (vivekgoe) [inline optimum/habana/accelerate/utils/transformer_engine.py:46]: Will be good to add a comment that when creating te.linear we are creating it without parameters and then shallow copying parameters from original nn.linear to te.linear layer.
Comment is good because this is different from what is being done for GPU in original code where layer is created with parameters but then nn.linear parameters are physically copies to te.linear.

File: optimum/habana/accelerate/utils/transformer_engine.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+            )
+            te_module.weight = module.weight
+
+            if has_bias:
+                te_module.bias = module.bias
+
+            setattr(model, name, te_module)
+        elif isinstance(module, te.Linear) and not to_transformer_engine and _convert_linear:
+            has_bias = module.bias is not None
```
Comment (vivekgoe) [inline optimum/habana/accelerate/utils/transformer_engine.py:55]: It will be good to check this case in a unit-test. Here we converting a te.Linear to nn.Linear

File: optimum/habana/accelerate/utils/transformer_engine.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+
+def has_transformer_engine_layers(model):
+    """
+    Returns whether a given model has some `transformer_engine` layer or not.
+    """
+    if not is_fp8_available():
+        raise ImportError("Using `has_transformer_engine_layers` requires transformer_engine to be installed.")
+    for m in model.modules():
+        if isinstance(m, (te.Linear)):
```
Comment (vivekgoe) [inline optimum/habana/accelerate/utils/transformer_engine.py:75]: Lets add support for SDPA also.

File: optimum/habana/accelerate/utils/transformer_engine.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+            return True
+    return False
+
+def convert_model(model):
+    if not has_transformer_engine_layers(model):
```
Comment (vivekgoe) [inline optimum/habana/accelerate/utils/transformer_engine.py:80]: Why is this function not there in https://github.com/huggingface/accelerate/blob/main/src/accelerate/utils/transformer_engine.py

File: optimum/habana/accelerate/utils/transformer_engine.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+        with torch.no_grad():
+            _convert_model(model)
+        model._converted_to_transformer_engine = True
+    return model
+
+def get_fp8_format(fp8_format):
+    if fp8_format == "E5M2":
```
Comment (vivekgoe) [inline optimum/habana/accelerate/utils/transformer_engine.py:87]: Why is this function not there in https://github.com/huggingface/accelerate/blob/main/src/accelerate/utils/transformer_engine.py

File: optimum/habana/accelerate/utils/transformer_engine.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+        return te.recipe.Format.E5M2
+    elif fp8_format == "HYBRID":
+        return te.recipe.Format.HYBRID
+    else:
+        raise ValueError
+
+def get_fp8_recipe(fp8_config):
+    fp8_config = dict(fp8_config) if fp8_config is not None else {}
```
Comment (vivekgoe) [inline optimum/habana/accelerate/utils/transformer_engine.py:95]: Why is this function not there in https://github.com/huggingface/accelerate/blob/main/src/accelerate/utils/transformer_engine.py

File: optimum/habana/peft/layer.py [hunk 9]
```diff
@@ -2,6 +2,8 @@
 
 import torch
 
+import habana_frameworks.torch.hpex.experimental.transformer_engine as te
+from peft.utils.other import transpose
 
 def GaudiLoraLayerLinearForward(self, x: torch.Tensor, *args: Any, **kwargs: Any) -> torch.Tensor:
```
Comment (vivekgoe) [inline optimum/habana/peft/layer.py:8]: If we end up creating a LoRALinear class, then we should unify this function with the new class.

File: optimum/habana/peft/layer.py [hunk 10]
```diff
@@ -29,3 +31,41 @@ def GaudiLoraLayerLinearForward(self, x: torch.Tensor, *args: Any, **kwargs: Any
 
     result = result.to(previous_dtype)
     return result
+
+class LoRALinear:
```
Comment (vivekgoe) [inline optimum/habana/peft/layer.py:35]: Why not derive LoRALinear from PEFTLinear and monkey patch PEFTLinear at init itself? That way you will not need replace_forward at all, class created will be of type LoRALinear to begin with.

File: optimum/habana/peft/layer.py [hunk 11]
```diff
@@ -29,3 +31,41 @@ def GaudiLoraLayerLinearForward(self, x: torch.Tensor, *args: Any, **kwargs: Any
...
+    def __init__(self, module):
+        has_bias = module.bias is not None
+        self.module = module
+        self.module.te_linear =  te.Linear(module.in_features, module.out_features, bias=has_bias, params_dtype=module.weight.dtype, skip_weight_param_allocation=True)
```
Comment (vivekgoe) [inline optimum/habana/peft/layer.py:39]: This got created without parameters, where do actual parameters get hooked to it?

**Review by vivekgoe — CHANGES_REQUESTED** [review]

File: optimum/habana/peft/layer.py (hunk 11 above)
Comment (vivekgoe) [reply optimum/habana/peft/layer.py:39]: Done in _linear function (Line 42).

File: optimum/habana/accelerate/utils/transformer_engine.py (hunk 2 above)
Comment (vivekgoe) [inline optimum/habana/accelerate/utils/transformer_engine.py:43]: Add condition to skip LoRA-A LoRA-B from TE wrapping. Does not give much performance gain and there is a chance of accuracy degrade.

File: optimum/habana/accelerate/utils/transformer_engine.py [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+    if "fp8_format" in fp8_config:
+        fp8_config['fp8_format'] = get_fp8_format(fp8_config['fp8_format'])
+    fp8_recipe_handler = te.recipe.DelayedScaling(**fp8_config)
+    fp8_recipe_handler.backend = "TE"
+    return fp8_recipe_handler
+
+class FP8ContextWrapper:
+    def __init__(self, ctx, fp8_recipe):
```
Comment (vivekgoe) [inline optimum/habana/accelerate/utils/transformer_engine.py:103]: Add comments explaining what class is doing.

File: optimum/habana/accelerate/utils/transformer_engine.py (hunk 1 above)
Comment (SanjuCSudhakaran) [reply optimum/habana/accelerate/utils/transformer_engine.py:36]: As discussed, will handle SDPA changes separately.

File: optimum/habana/accelerate/utils/transformer_engine.py (hunk 3 above)
Comment (SanjuCSudhakaran) [reply optimum/habana/accelerate/utils/transformer_engine.py:43]: Done

File: optimum/habana/accelerate/utils/transformer_engine.py (hunk 5 above)
Comment (SanjuCSudhakaran) [reply optimum/habana/accelerate/utils/transformer_engine.py:72]: As discussed, will handle SDPA changes separately.

File: optimum/habana/accelerate/utils/transformer_engine.py (hunk 2 above)
Comment (SanjuCSudhakaran) [reply optimum/habana/accelerate/utils/transformer_engine.py]: Latest peft doesn't require this special handling. so these changes are not required anymore since we moved to peft0.10.0

File: optimum/habana/accelerate/utils/transformer_engine.py (hunk 6 above)
Comment (SanjuCSudhakaran) [reply optimum/habana/accelerate/utils/transformer_engine.py:81]: The same sequence of operations is required in multiple locations in our implementations. So this function acts as wrapper.

File: optimum/habana/accelerate/utils/transformer_engine.py (hunk 8 above)
Comment (SanjuCSudhakaran) [reply optimum/habana/accelerate/utils/transformer_engine.py]: The same sequence of operations is required in multiple locations in our implementations. So this function acts as wrapper.

File: optimum/habana/accelerate/utils/transformer_engine.py (hunk 7 above)
Comment (SanjuCSudhakaran) [reply optimum/habana/accelerate/utils/transformer_engine.py]: Handled the functionality in `get_fp8_recipe` in a different way.

File: optimum/habana/peft/layer.py (hunk 9 above)
Comment (SanjuCSudhakaran) [reply optimum/habana/peft/layer.py:5]: Latest peft doesn't require this special handling. so these changes are not required anymore since we moved to peft0.10.0

File: optimum/habana/peft/layer.py (hunk 10 above)
Comment (SanjuCSudhakaran) [reply optimum/habana/peft/layer.py]: Latest peft doesn't require this special handling. so these changes are not required anymore since we moved to peft0.10.0

File: optimum/habana/accelerate/utils/transformer_engine.py (hunk 12 above)
Comment (SanjuCSudhakaran) [reply optimum/habana/accelerate/utils/transformer_engine.py:103]: Done

File: optimum/habana/accelerate/accelerator.py [hunk 13]
```diff
@@ -220,6 +221,11 @@ def __init__(
             **kwargs,
         )
 
+        if self.state.is_fp8_enabled:
+            if self.fp8_recipe_handler is None:
+                self.fp8_recipe_handler = GaudiFP8RecipeKwargs()
```
Comment (vivekgoe) [inline optimum/habana/accelerate/accelerator.py:226]: Why do we need this init? Shouldn't this be already initialized in  L205 above.

File: optimum/habana/accelerate/accelerator.py [hunk 14]
```diff
@@ -220,6 +221,11 @@ def __init__(
...
+            self.fp8_recipe_handler = get_fp8_recipe(self.fp8_recipe_handler)
```
Comment (vivekgoe) [inline optimum/habana/accelerate/accelerator.py:227]: In original accelerator.py code in this function is called from prepare_model() function. Please add a comment that the reason you moved this to __init__() is to cover both paths with and without deepspeed.

File: optimum/habana/accelerate/accelerator.py [hunk 15]
```diff
@@ -672,6 +657,7 @@ def _prepare_deepspeed(self, *args):
                     result[i] = scheduler
             # pointing for deepspeed_engine_wrapped.backward()
             self.deepspeed_engine_wrapped = DeepSpeedEngineWrapper(engine)
+
```
Comment (vivekgoe) [inline optimum/habana/accelerate/accelerator.py:660]: Delete empty line.

File: optimum/habana/accelerate/state.py [hunk 16]
```diff
@@ -173,6 +173,7 @@ def __init__(
                 if mixed_precision is None
                 else mixed_precision.lower()
             )
+            self.is_fp8_enabled = mixed_precision == "fp8"
```
Comment (vivekgoe) [inline optimum/habana/accelerate/state.py:176]: It is risky to set mixed_precision to "fp8", since it is being used at many other places. All we want to do is to set is_fp8_enabled = True here, Can we not do this in GaudiAccelerator __Init__ itself? Let's discuss this further.

File: optimum/habana/accelerate/utils/transformer_engine.py [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+
+    def __exit__(self, exc_type, exc_value, exc_traceback):
+        self.fp8_ctx.__exit__(exc_type, exc_value, exc_traceback)
+        self.ctx.__exit__(exc_type, exc_value, exc_traceback)
+
+    @staticmethod
+    def create_fp8_context(fp8_recipe):
+        return te.fp8_autocast(enabled=True, fp8_recipe=fp8_recipe)
+
+    # `_gradient_checkpointing_func` always takes the function to be recomputed as the first argument. The function
```
Comment (vivekgoe) [inline optimum/habana/accelerate/utils/transformer_engine.py:119]: write this as a standard python help string with ''' my comments ''' within function.

File: optimum/habana/accelerate/utils/transformer_engine.py [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+    # below wraps this first argument with `transformer_engine` `activation_checkpointing` context.
+    @staticmethod
+    def _gradient_checkpointing_wrap(func, *args, **kwargs):
+        _args = list(args)
+        _args[0] = activation_checkpointing()(_args[0])
+        args = tuple(_args)
+
+        return func(*args, **kwargs)
+
+    @staticmethod
```
Comment (vivekgoe) [inline optimum/habana/accelerate/utils/transformer_engine.py:129]: Short comment on what this function is doing.

File: optimum/habana/transformers/training_args.py [hunk 19]
```diff
@@ -487,7 +492,9 @@ def __post_init__(self):
 
         # if training args is specified, it will override the one specified in the accelerate config
         mixed_precision_dtype = os.environ.get("ACCELERATE_MIXED_PRECISION", "no")
-        if self.bf16:
+        if self.fp8:
+            mixed_precision_dtype = "fp8"
```
Comment (vivekgoe) [inline optimum/habana/transformers/training_args.py:496]: It is risky to overwrite mixed_precision_dtype to "fp8", let's figure out another way of doing this.

File: optimum/habana/transformers/trainer.py [hunk 20]
```diff
@@ -673,6 +673,10 @@ def hpu_deepspeed_checkpointing(function, *checkpoint_args):
                 transformers.modeling_utils.checkpoint = lazy_mode_checkpointing
 
             self.model.gradient_checkpointing_enable(gradient_checkpointing_kwargs=gradient_checkpointing_kwargs)
+
+            # Wrap `_gradient_checkpointing_func` in the model with `transformer_engine` `activation_checkpointing` context.
```
Comment (vivekgoe) [inline optimum/habana/transformers/trainer.py:677]: Add short line on what is activation_checkpointing context and why is it needed here.

**Review by vivekgoe — CHANGES_REQUESTED** [review]
We also need to update examples/language_modeling/README file with example command for fp8 training.

File: optimum/habana/accelerate/accelerator.py (hunk 14 above)
Comment (SanjuCSudhakaran) [reply optimum/habana/accelerate/accelerator.py:229]: Done

File: optimum/habana/accelerate/accelerator.py (hunk 15 above)
Comment (SanjuCSudhakaran) [reply optimum/habana/accelerate/accelerator.py]: Done

File: optimum/habana/accelerate/utils/transformer_engine.py (hunk 17 above)
Comment (SanjuCSudhakaran) [reply optimum/habana/accelerate/utils/transformer_engine.py]: Done

File: optimum/habana/accelerate/utils/transformer_engine.py (hunk 18 above)
Comment (SanjuCSudhakaran) [reply optimum/habana/accelerate/utils/transformer_engine.py:131]: Done

File: optimum/habana/transformers/trainer.py (hunk 20 above)
Comment (SanjuCSudhakaran) [reply optimum/habana/transformers/trainer.py:677]: Done

File: optimum/habana/accelerate/accelerator.py (hunk 13 above)
Comment (SanjuCSudhakaran) [reply optimum/habana/accelerate/accelerator.py:226]: Setting with default values if user doesn't specify anything.

File: optimum/habana/accelerate/state.py (hunk 16 above)
Comment (SanjuCSudhakaran) [reply optimum/habana/accelerate/state.py:176]: As discussed, we will proceed with this implementation.

File: optimum/habana/transformers/training_args.py (hunk 19 above)
Comment (SanjuCSudhakaran) [reply optimum/habana/transformers/training_args.py:496]: As discussed, we will proceed with this implementation.

File: optimum/habana/accelerate/utils/transformer_engine.py [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+            if has_bias:
+                te_module.bias = module.bias
+
+            setattr(model, name, te_module)
+        elif isinstance(module, te.Linear) and not to_transformer_engine and _convert_linear:
+            has_bias = module.bias is not None
+            new_module = torch.nn.Linear(
+                module.in_features, module.out_features, bias=has_bias, dtype=module.weight.dtype, device=module.weight.device
+            )
+            new_module.weight.copy_(module.weight)
```
Comment (bhargaveede) [inline optimum/habana/accelerate/utils/transformer_engine.py:56]: why are we copying here and it's not the same when converting from nn to te.

**Review by vivekgoe — APPROVED** [review]

**Comment by astachowiczhabana** [comment]
https://github.com/huggingface/optimum-habana/pull/1096/files
